File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.c
int nCr(int n, int r, int k_cap) {
    if (r < 0 || r > n) return 0;
    if (r == 0 || r == n) return 1;
    if (r > n - r) {
        r = n - r;
    }

    long long res = 1;
    for (int i = 1; i <= r; i++) {
        res = res * (n - i + 1) / i; // Safe to multiply using long long
        
        // Direct, accurate capping without truncation errors
        if (res >= k_cap) {
            return k_cap;
        }
    }
    return (int)res;
}


int arrangements(int* counts, int k_cap) {
    int total = 0;
    for (int i = 0; i < 26; i++) {
        total += counts[i];
    }

    long long ways = 1;
    for (int i = 0; i < 26; i++) {
        if (counts[i] > 0) {
            int combinations = nCr(total, counts[i], k_cap);
            if (combinations == 0) return 0;
            
            ways *= combinations;
            // Direct cap validation
            if (ways >= k_cap) {
                return k_cap;
            }
            total -= counts[i];
        }
    }
    return (int)ways;
}


char* smallestPalindrome(char* s, int k) {
    int len = strlen(s);
    int counts[26] = {0};
    char mid_char = '\0';

    for(int i = 0;i < len; i++){
        counts[s[i] - 'a']++;
    }

    int half_counts[26] = {0};
    int total_len = 0;
    for(int i = 0; i < 26; i++){
        half_counts[i] = counts[i] / 2;
        total_len += half_counts[i];

        if(counts[i] % 2 != 0){
            mid_char = 'a' + i;
        }
    }

    if(arrangements(half_counts , k) < k){
        char* empty_res = (char*)malloc(sizeof(char));
        empty_res[0] = '\0';
        return empty_res;
    }

    char* left_half = (char*)malloc((total_len + 1)* sizeof(char));
    int idx = 0;
    for(int step = 0; step < total_len; step++){
        for(int i = 0; i < 26; i++){
            if(half_counts[i] == 0) continue;
            half_counts[i]--;
            int w = arrangements(half_counts , k);
            if(w >= k){
                left_half[idx++] = 'a' + i;
                break;
            }else{
                k -= w;
                half_counts[i]++;
            }
        }
    }
    left_half[idx] = '\0';

    char* result = (char*)malloc((len + 1) * sizeof(char));
    int r_idx = 0;
    for(int i = 0; i < total_len; i++){
        result[r_idx++] = left_half[i];
    }

    if(mid_char != '\0'){
        result[r_idx++] = mid_char;
    }

    for(int i = total_len - 1; i >= 0; i--){
        result[r_idx++] = left_half[i];
    }

    result[r_idx] = '\0';
    free(left_half);
    return result;
}

File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.py
from collections import Counter
import math

class Solution:
    def smallestPalindrome(self, s: str, k: int) -> str:
        # Step 1: Count character frequencies
        counts = Counter(s)
        
        # Step 2: Initialize arrays for the left half and trace any middle character
        half_counts = [0] * 26
        mid_char = ""
        total_len = 0
        
        for char, count in counts.items():
            idx = ord(char) - ord('a')
            half_counts[idx] = count // 2
            total_len += half_counts[idx]
            
            # An odd frequency means this character sits in the exact middle
            if count % 2 != 0:
                mid_char = char
        
        # Helper function to compute total unique arrangements capped at k
        def count_arrangements(pool_counts: list[int]) -> int:
            total = sum(pool_counts)
            ways = 1
            for current_count in pool_counts:
                if current_count > 0:
                    ways *= math.comb(total, current_count)
                    if ways >= k:
                        return k  # Cap early to save memory and processing time
                    total -= current_count
            return ways

        # Step 3: Fast feasibility check
        if count_arrangements(half_counts) < k:
            return ""

        # Step 4: Greedily build the left half character by character
        left_half = []
        for step in range(total_len):
            for i in range(26):
                if half_counts[i] == 0:
                    continue
                
                # Speculatively take one character
                half_counts[i] -= 1
                arrangements = count_arrangements(half_counts)
                
                # Check if our target k falls within this branch
                if arrangements >= k:
                    left_half.append(chr(ord('a') + i))
                    break  # Lock character and move to next position
                else:
                    k -= arrangements  # Skip branch and reduce k
                    half_counts[i] += 1  # Put the character back

        # Step 5: Mirror and assemble the final palindrome
        left_str = "".join(left_half)
        return left_str + mid_char + left_str[::-1]
